File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static boolean[] visited;
    static HashMap<Integer, ArrayList<Integer>> graph;
    static int n;
    static int cnt = -1;
    static int x;
    static int y;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        x = Integer.parseInt(st.nextToken());
        y = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(br.readLine());
        graph = new HashMap<>();
        visited = new boolean[n+1];
        for(int i = 1; i < n + 1; i++){
            graph.put(i, new ArrayList<>());
        }
        for(int i = 0; i < m; i++){
            st = new StringTokenizer(br.readLine());
            int v = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            graph.get(v).add(e);
            graph.get(e).add(v);
        }
        DFS(x, 0);
        System.out.println(cnt);
        br.close();
    }

    private static void DFS(int r, int num){
        visited[r] = true;
        if(r == y){
            cnt = num;
        }
        for(int i : graph.get(r)){
            if(!visited[i]){
                DFS(i, num + 1);
            }
        }
    }
}

File: 백준/Silver/2644. 촌수계산/촌수계산.py
n = int(input())
a, b = map(int, input().split())   
m = int(input())
graph = [[] for _ in range(n+1)]
visited = [0] * (n+1)

for i in range(m):
    v, e = map(int, input().split())
    graph[v].append(e)
    graph[e].append(v)



def dfs(v, depth):
    if visited[v] != 0:
        return
    
    visited[v] += depth

    for i in graph[v]:
        if visited[i] == 0:
            dfs(i, depth+1)


dfs(a, 1)


if(visited[b] == 0):
    print(-1)
else:
    print(visited[b] - 1)